src/CMakeLists.txt

```diff
@@ -628,31 +628,40 @@ install(TARGETS ceph-mon DESTINATION bin)
 
 # OSD/ObjectStore
 # make rocksdb statically
-if(NOT ALLOCATOR STREQUAL "jemalloc")
-  set(disable_jemalloc "DISABLE_JEMALLOC=1")
-endif()
 
-set(ROCKSDB_EXTRA_CXXFLAG -fPIC -Wno-unused-variable)
-if(NOT CMAKE_BUILD_TYPE STREQUAL Debug)
-  set(ROCKSDB_EXTRA_CXXFLAG ${ROCKSDB_EXTRA_CXXFLAG} -DNDEBUG)
-endif(NOT CMAKE_BUILD_TYPE STREQUAL Debug)
+set(ROCKSDB_CXX_FLAGS "-DCMAKE_CXX_FLAGS=-fPIC")
+
+if(ALLOCATOR STREQUAL "jemalloc")
+  set(ROCKSDB_JEMALLOC "-DWITH_JEMALLOC=ON")
+endif()
 
-set(ROCKSDB_CXX "${CMAKE_CXX_COMPILER}")
 if (WITH_CCACHE AND CCACHE_FOUND)
-  set(ROCKSDB_CXX "ccache" "${CMAKE_CXX_COMPILER}")
+	  set(ROCKSDB_CXX "-DCMAKE_CXX_COMPILER=ccache")
+    set(ROCKSDB_CXX_ARG1 "-DCMAKE_CXX_COMPILER_ARG1=${CMAKE_CXX_COMPILER}")
+else(WITH_CCACHE AND CCACHE_FOUND)
+	  set(ROCKSDB_CXX "-DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}")
 endif(WITH_CCACHE AND CCACHE_FOUND)
 
-add_custom_target(build_rocksdb
-    COMMAND
-    PORTABLE=1 ${disable_jemalloc} $(MAKE) static_lib CXX="${ROCKSDB_CXX}" EXTRA_CXXFLAGS="${ROCKSDB_EXTRA_CXXFLAG}"
-    WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}/src/rocksdb
-    COMMENT "rocksdb building")
+list(APPEND ROCKSDB_AR -DCMAKE_AR=${CMAKE_AR})
+
+# we use an external project and copy the sources to bin directory to ensure
+# that object files are built outside of the source tree.
+include(ExternalProject)
+ExternalProject_Add(rocksdb_ext
+    SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/rocksdb
+    CMAKE_ARGS ${ROCKSDB_CXX} ${ROCKSDB_CXX_ARG1} ${ROCKSDB_AR} ${ROCKSDB_JEMALLOC} ${ROCKSDB_CXX_FLAGS}
+    BINARY_DIR ${CMAKE_CURRENT_BINARY_DIR}/rocksdb
+    BUILD_COMMAND $(MAKE) rocksdblib
+    INSTALL_COMMAND "true")
+
+# force rocksdb make to be called on each time
+ExternalProject_Add_Step(rocksdb_ext forcebuild DEPENDEES configure DEPENDERS COMMAND "true" build ALWAYS 1)
```

Comment:
with following patch on top of yours my cmake does not complain
```diff

-    INSTALL_COMMAND "true")
+  SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/rocksdb
+  CMAKE_ARGS ${ROCKSDB_CXX} ${ROCKSDB_CXX_ARG1} ${ROCKSDB_AR} ${ROCKSDB_JEMALLOC} ${ROCKSDB_CXX_FLAGS}
+  BINARY_DIR ${CMAKE_CURRENT_BINARY_DIR}/rocksdb
+  BUILD_COMMAND $(MAKE) rocksdblib
+  INSTALL_COMMAND "true")
```

i will push the change to gitbuilder to verify your/my theory.